swift/ql/lib/codeql/swift/elements/expr/ArithmeticOperation.qll

```diff
@@ -0,0 +1,99 @@
+private import codeql.swift.elements.expr.Expr
+private import codeql.swift.elements.expr.BinaryExpr
+private import codeql.swift.elements.expr.PrefixUnaryExpr
+private import codeql.swift.elements.expr.DotSyntaxCallExpr
+
+/**
+ * An arithmetic operation, such as:
+ * ```
+ * a + b
+ * ```
+ */
+class ArithmeticOperation extends Expr {
+  ArithmeticOperation() {
+    this instanceof BinaryArithmeticOperation or
+    this instanceof UnaryArithmeticOperation
+  }
+
+  Expr getAnOperand() {
+    result = this.(BinaryArithmeticOperation).getAnOperand()
+    or
+    result = this.(UnaryArithmeticOperation).getOperand()
+  }
+}
+
+/**
+ * A binary arithmetic operation, such as:
+ * ```
+ * a + b
+ * ```
+ */
+abstract class BinaryArithmeticOperation extends BinaryExpr { }
```

Comment:
We shouldn't create public abstract AST classes as this creates a dangerous footgun for QL writers. If someone were to create a class that extends `BinaryArithmeticOperation` in a query it'll cause everything that depends on the AST layer to be recomputed.

We either need to make this class `private` (which doesn't seem to be the right thing to do), or list its subclasses in the charpred.

Reply:
Yes, I noticed you'd done something different in `LogicalOperation.qll` and I was meaning to bring up the tradeoff... I don't *like* the way `LogicalAndExpr` doesn't extend `BinaryLogicalOperation`.  I had thought this a problem, but the more I think about it the more I'm not sure it actually is.

I guess I'm talking myself around to adopting your design...

Reply:
Done.